docs/nebula-design-concepts.md

```diff
@@ -0,0 +1,211 @@
+
+- Add Raft Learner
+
+      When a new server joins the cluster, it can be added as a learner, which can neither vote nor count towards quorum. Once it catches up to the leader's logs, it can be promoted to follower as a normal voting node.
+
+- Load-balance
+
+      Migrating the partitions on an overworked server to other relatively idle servers to increases availability and capacity of the system.
+
+
+
+![image](https://user-images.githubusercontent.com/42762957/61120371-6f0fc200-a4cf-11e9-8c41-9e531380205b.png)
+
+
+
+### Metadata service
+
+The process corresponding to the meta service is nebula-metad, and its main functions are:
+
+-  User management
+
+      In Nebula Graph different roles are assigned diverse privileges. We provide the following native roles: God user, Admin, User and Guest.
+
+- Cluster configuration management
+
+      Metadata service manages the servers and partitions in the cluster, e.g. records location of the partitions, receives heartbeat from servers, etc. It balances the partitions and manages the communication traffic in case of server failure.
+
+- Graph space management
+
+      Metadata service stores the metadata of all spaces in the cluster and tracks changes that take place in these spaces, like adding, dropping space, modifying graph space configuration (Raft copies).
+
+-   Schema management
+
+      Nebula Graph is a strong typed database.
+
+      - Types of tag and edge properties are recorded by meta service. Supported data types are: int, double, timestamp, list, etc.
+
+      - Multi-version management, supporting adding, modifying and deleting schema, and recording its version.
+
+      - TTL (time-to-live) management, supporting automatic data deletion and space reclamation.
+
+The meta service is stateful, and just like the storage service, it persists data to a key-value store.
+
+
+![image](https://user-images.githubusercontent.com/42762957/61120413-8cdd2700-a4cf-11e9-8846-14b5d8bd6693.png)
+
+
+### Query Engine & Query Language (nGQL)
+
+The process corresponding to the query engine is nebula-graphd, which consists
+of nodes that are stateless, equally privileged and isolated from each other.
+
+Query engine accepts the message from the client and generates the query plan after lexical analysis (Lexer), parsing (Parser) and query optimization. Then the execution plan will be passed to the execution engine. The query execution engine takes the query plans and interacts with meta server and the storage engine to retrieve the schema and data.
+
+The main optimizations of the query engine are:
+
+- Asynchronous and parallel execution
+
+      I/O operations and network transmission are time-consuming. Thus asynchronous and parallel operations are adopted in Nebula query engine to guarantee quality of service (QoS).Also, a separate resource pool is set for each query to avoid the long tail effect of those time-consuming queries.
```

Comment:
I/O operations and network transmission are time-consuming. Thus asynchronous and parallel operations are widely adopted in the query engine to reduce the latency and to improve the overall throughput.